### ISSUE TITLE
Query Parameters shouldn't be mandatory?

### ISSUE THREAD

**Issue by Andersmholmgren** [google/uri.dart#14]
It seems unintuitive to fail a match when query parameters are missing.

e.g. a route like

```
search{?name}
```

curl -X GET '/search' would fail to match but that is not what I would consider typical behaviour. You normally want the request to match and the `name` parameter to be null.

**Comment by justinfagnani** [comment]
Picking this up again, I think the best option is to add a parameter to `UriPattern.match()`, and possibly remove `UriParser.parse()`. When query parameters are optional, `UriMatch` can have a field which says whether all of the parameters had matches. Between that and the original URI being available as `UriMatch.input` users should be able to customize matching with specific optional and required parameters by wrapping.

This approach should be forward compatible if we find a way to make parameters optional/required on an individual basis in the template.

**Comment by Andersmholmgren** [comment]
That sounds good to me. 

Another option to a bool saying if all query params matched you could expose an Iterable of unmatched query params, or of those that did match (depending on your preference).

There is already a Map called parameters that seems to relate to this. Not sure how that would fit

**Comment by justinfagnani** [comment]
`UriPattern` could expose the set of parameters it's matching on, then finding unmatched parameters is just a set difference.

**Comment by justinfagnani** [comment]
One detail is to work out the exact semantics of this new parameter. Does it make all parameters optional, or just query parameters? Path parameters probably shouldn't be optional, but I have seen some schemes where key/value pairs are encoded in paths and fragments. Right now `UriPattern` and `UriMatch` do not specify anything regarding different types of parameters.

Thinking this through could lead to another solution: a constructor parameter on `UriParser`, leaving the strictness of `match()` up to the implementation and even instance in our case.

**Comment by Andersmholmgren** [comment]
Sounds gtm

**Comment by justinfagnani** [comment]
@Andersmholmgren any chance you can do the PR for this?

**Comment by Andersmholmgren** [comment]
I think if there is just the one bool parameter then it should only apply to query params. If we want to support optional path params (which I don't think there is a major case for) then it needs to be kept separate. Either another bool param or the param needs to be more like an enum

**Comment by Andersmholmgren** [comment]
haha yeah I had already resigned to doing it myself although I must admit you "picking this up again" had me all excited ;-)

**Comment by justinfagnani** [comment]
This goes to why a constructor param on `UriParser` might be better - `match()` wouldn't say anything about strictness, each `UriPattern` instance would in it's own way, and their APIs can evolve on their own. I'd like to add a Rails-style pattern for instance, and I'm not even sure they support less-strict matching.

**Comment by justinfagnani** [comment]
Yeah, sorry about that. I've moved to Polymer full-time, and while I'd like to have time to maintain all my Dart packages, I just don't have enough to do it right. I can certainly help here and there and do reviews though!

**Comment by Andersmholmgren** [comment]
Actually yes I see what you mean by a ctr to `UriParser`. I like that. And the result `UriMatch` object simply provides details of what did match for those that care.

I think that is the cleaner option and will approach it that way. thanks

Ah time. Something we could all do with more of ;-)

**Comment by justinfagnani** [comment]
Awesome! Let me know if you need any help, I know the guts of matching are a little uglier than I would like. I can always look at a branch.

**Comment by Andersmholmgren** [comment]
fyi about to start dev on this now

**Comment by sethladd** [comment]
How's the patch coming along?

**Comment by Andersmholmgren** [comment]
https://github.com/google/uri.dart/pull/23

**Comment by JKRhb** [comment]
With the merge of #23, I think this issue was resolved, right?
